### PR TITLE
tracking: Lambda as child processes

### PR DESCRIPTION
**Pull request by dominictarr**
Each lambda runs as a child process.

This is a tracking PR. I've made a pretty heavy handed approach to restructuring things and removing old features.

For example, there was previously a single global WorkerPool, but that increases complexity, and possible resource leaks, and doesn't seem to any advantages that I could think of.

In the original, each set of routes had a handler, but now, there is just one handler - I havn't removed the handler yet but that seems like the next step. The worker pool won't need a reference to the handler, it can just return a promise representing the request.

This also doesn't have a way to do handle requests one at a time yet.

I think the simplest would be to a) just create a new short lived worker for each request, or b) wait for the current worker to finish, or c) create new workers at any time but kill them after a 30 seconds (or they respond)

most of the code in the worker is still the same, because I didn't need to change it to make this work.

### CODE REVIEW

**Comment by dominictarr** [comment]
tests pass, but typechecks don't

**Comment by Raynos** [comment]
@dominictarr I believe we wish to rewrite this module effectively.

Can we create a new repository / name under https://github.com/orgs/optoolco/repositories ;

You can push the git history of this repo to it and open the same PR over there and then we can close this one.

Once the rewrite is done I can archive this repository.

**Comment by Raynos** [comment]
The PR itself seems reasonable, I will add github actions to run tests once we moved the repository.

**Comment by Raynos** [comment]
I updated `master` with the necessary github actions changes to run CI.

You can `git rebase origin/master` and push to get CI running for this pull request.

File: package.json [hunk 1]
```diff
@@ -64,5 +65,9 @@
       "type": "MIT",
       "url": "http://github.com/Raynos/fake-ses/raw/master/LICENSE"
     }
-  ]
+  ],
+  "dependencies": {
+    "mkdirp": "^0.5.5",
```
Comment (Raynos) [inline package.json:70]: nodejs has `fs.mkdir(folder, { recursive: true })` so this dependency is not needed.

File: package.json [hunk 2]
```diff
@@ -64,5 +65,9 @@
...
+    "ncp": "^2.0.0"
```
Comment (Raynos) [inline package.json:71]: There is no recursive copy in node core which is annoying af, so `ncp` is still needed :(

File: docker.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
+var path = require('path')
+var ncp = require('ncp')
+var mkdirp = require('mkdirp')
+var fs = require('fs')
+var cp = require('child_process')
+var fetch = require('node-fetch')
+//run lambda process within docker.
+
+function copy (src, dest, cb) {
```
Comment (Raynos) [inline docker.js:9]: nitpick: we try to consistently use async/await in all our code and avoid callbacks. can discuss in the code session / call tonight

File: child-process-worker.js [hunk 4]
```diff
@@ -1,13 +1,17 @@
 const path = require('path')
 const childProcess = require('child_process')
 
-const WORKER_PATH = path.join(__dirname, 'worker.js')
+var WORKER_PATH = '/tmp/worker.js' //path.join(__dirname, 'worker.js')
+
+const WorkerMain = require('./worker')
+require('fs').writeFileSync('/tmp/worker.js', ';('+WorkerMain.toString()+')();function __name (){}; ')
```
Comment (dominictarr) [inline child-process-worker.js:7]: note the `__name` noop is needed because that function is inserted by the esbuild.

**Comment by dominictarr** [comment]
because fake-api-gateway-lambda might be used by a process that has been bundled by esbuild, it was necessary to change the approach of running the worker. accessing the file directly via __dirname is possible, because the __dirname is changed, and the file isn't really there. it has no deps, so I wrap it in a function, get the source string, write it to a file in tmp, then run that file. esbuild inserts `__name(fn, string)` to enforce (?) the name of functions? anyway easyiest to insert a noop so that doesn't error.

**Comment by Raynos** [comment]
Can we delete the child process worker now that the docker worker is implemented ?

**Pull request merged**